src/Build/Construction/Solution/SolutionFile.cs

```diff
@@ -195,7 +200,39 @@ internal string FullPath
             {
                 // Should already be canonicalized to a full path
                 ErrorUtilities.VerifyThrowInternalRooted(value);
-                _solutionFile = value;
+                if (FileUtilities.IsSolutionFilterFilename(value))
+                {
+                    using (JsonDocument text = JsonDocument.Parse(File.ReadAllText(value)))
+                    {
+                        try
+                        {
+                            JsonElement solution = text.RootElement.GetProperty("solution");
+                            _solutionFile = solution.GetProperty("path").GetString();
+                            _solutionFilter = new HashSet<string>(NativeMethodsShared.IsLinux ? StringComparer.Ordinal : StringComparer.OrdinalIgnoreCase);
+                            foreach (JsonElement project in solution.GetProperty("projects").EnumerateArray())
+                            {
+                                _solutionFilter.Add(project.GetString());
+                            }
+                        }
+                        catch (Exception)
+                        {
+                            ProjectFileErrorUtilities.VerifyThrowInvalidProjectFile
+                                (
+                                    false /* just throw the exception */,
```

Comment:
Can we preserve information about the exception to pass up to the user?

Reply:
Does this not preserve information—in the form of an exception the user can catch and inspect?

Reply:
No, because a primary use case for this functionality is using `msbuild.exe foo.slnf` on the command line, where you get only MSBuild errors (not exceptions).